File: api/services/weighting.py
"""Provider weighting algorithms."""
import numpy as np
from typing import Dict, List
import logging
from config import settings

logger = logging.getLogger(__name__)


def softmax_weights(brier_scores: Dict[int, float], temperature: float = 1.0) -> Dict[int, float]:
    """
    Calculate provider weights using softmax on negative Brier scores.

    Args:
        brier_scores: Dictionary mapping provider_id to Brier score
        temperature: Temperature parameter for softmax (higher = more uniform)

    Returns:
        Dictionary mapping provider_id to weight
    """
    if not brier_scores:
        return {}

    provider_ids = list(brier_scores.keys())
    scores = np.array([brier_scores[pid] for pid in provider_ids])

    # Use negative scores (lower Brier is better)
    neg_scores = -scores / temperature

    # Softmax
    exp_scores = np.exp(neg_scores - np.max(neg_scores))  # Subtract max for numerical stability
    weights = exp_scores / exp_scores.sum()

    return {pid: float(w) for pid, w in zip(provider_ids, weights)}


def inverse_score_weights(brier_scores: Dict[int, float]) -> Dict[int, float]:
    """
    Calculate provider weights using inverse Brier scores.

    Args:
        brier_scores: Dictionary mapping provider_id to Brier score

    Returns:
        Dictionary mapping provider_id to weight
    """
    if not brier_scores:
        return {}

    provider_ids = list(brier_scores.keys())
    scores = np.array([brier_scores[pid] for pid in provider_ids])

    # Inverse weights (lower score = higher weight)
    epsilon = 1e-8
    inverse_scores = 1.0 / (scores + epsilon)

    # Normalize
    weights = inverse_scores / inverse_scores.sum()

    return {pid: float(w) for pid, w in zip(provider_ids, weights)}


def apply_weight_constraints(
    weights: Dict[int, float],
    sample_sizes: Dict[int, int],
    min_samples: int = None,
    weight_floor: float = None,
    weight_ceiling: float = None,
) -> Dict[int, float]:
    """
    Apply constraints to provider weights.

    Args:
        weights: Dictionary mapping provider_id to weight
        sample_sizes: Dictionary mapping provider_id to sample size
        min_samples: Minimum samples required for full weight (from settings if None)
        weight_floor: Minimum weight per provider (from settings if None)
        weight_ceiling: Maximum weight per provider (from settings if None)

    Returns:
        Constrained weights dictionary
    """
    if not weights:
        return {}

    if min_samples is None:
        min_samples = settings.min_samples_for_weight
    if weight_floor is None:
        weight_floor = settings.weight_floor
    if weight_ceiling is None:
        weight_ceiling = settings.weight_ceiling

    constrained = {}

    for provider_id, weight in weights.items():
        # Apply sample size constraint (linear ramp from 0 to min_samples)
        sample_size = sample_sizes.get(provider_id, 0)
        if sample_size < min_samples:
            sample_factor = sample_size / min_samples
            weight = weight * sample_factor

        # Apply floor and ceiling
        weight = max(weight_floor, min(weight_ceiling, weight))

        constrained[provider_id] = weight

    # Re-normalize to sum to 1.0
    total = sum(constrained.values())
    if total > 0:
        constrained = {pid: w / total for pid, w in constrained.items()}

    return constrained


def calculate_provider_weights(
    performances: List[Dict],
    method: str = "softmax",
    temperature: float = 1.0,
) -> Dict[int, float]:
    """
    Calculate provider weights from performance data.

    Args:
        performances: List of dicts with 'provider_id', 'brier_score', 'sample_size'
        method: Weighting method ('softmax' or 'inverse')
        temperature: Temperature for softmax method

    Returns:
        Dictionary mapping provider_id to weight
    """
    if not performances:
        logger.warning("No performance data provided for weighting")
        return {}

    # Extract data
    brier_scores = {p["provider_id"]: p["brier_score"] for p in performances}
    sample_sizes = {p["provider_id"]: p["sample_size"] for p in performances}

    # Calculate base weights
    if method == "softmax":
        weights = softmax_weights(brier_scores, temperature=temperature)
    elif method == "inverse":
        weights = inverse_score_weights(brier_scores)
    else:
        raise ValueError(f"Unknown weighting method: {method}")

    # Apply constraints
    weights = apply_weight_constraints(weights, sample_sizes)

    return weights


def equal_weights(provider_ids: List[int]) -> Dict[int, float]:
    """
    Calculate equal weights for all providers.

    Args:
        provider_ids: List of provider IDs

    Returns:
        Dictionary mapping provider_id to equal weight
    """
    if not provider_ids:
        return {}

    weight = 1.0 / len(provider_ids)
    return {pid: weight for pid in provider_ids}

File: web/src/lib/odds/weighting.ts
/**
 * Provider weighting logic
 * Simplified for serverless use without historical performance tracking
 */

import { WeightMap, ProviderAccuracyRow } from './types';

/**
 * Get provider weights for a given league and market type
 * Simplified version - uses equal weights for all providers
 * In a full implementation, this would use historical Brier scores
 *
 * @param league League code (e.g., 'AFL', 'NRL')
 * @param marketType Market type (e.g., 'moneyline_2way', 'moneyline_3way')
 * @returns Weight map for providers
 */
export function getProviderWeights(
  league: string,
  marketType: string
): WeightMap {
  // For now, return equal weights for all bookmakers
  // In future, this could be enhanced with:
  // - Historical performance data stored in a database
  // - Dynamic weight calculation based on Brier scores
  // - League-specific and market-specific weighting

  // Equal weight strategy (simplest and safest for MVP)
  return {
    // All bookmakers get equal weight of 1.0
    // When aggregating, we'll use all available bookmaker odds
    default: 1.0,
  };
}

/**
 * Generate weight map for a list of provider IDs
 * @param providerIds List of provider/bookmaker identifiers
 * @param league League code
 * @param marketType Market type
 * @returns Weight map with equal weights for all providers
 */
export function generateWeightMapForProviders(
  providerIds: string[],
  league: string = 'default',
  marketType: string = 'default'
): WeightMap {
  const weights: WeightMap = {};

  // Assign equal weight to each provider
  const equalWeight = 1.0 / providerIds.length;

  for (const providerId of providerIds) {
    weights[providerId] = equalWeight;
  }

  return weights;
}

/**
 * Normalize weights to sum to 1.0
 * @param weights Weight map
 * @returns Normalized weight map
 */
export function normalizeWeights(weights: WeightMap): WeightMap {
  const total = Object.values(weights).reduce((sum, w) => sum + w, 0);

  if (total === 0) {
    throw new Error('Total weight cannot be zero');
  }

  const normalized: WeightMap = {};
  for (const [provider, weight] of Object.entries(weights)) {
    normalized[provider] = weight / total;
  }

  return normalized;
}

/**
 * Apply weight constraints (floor and ceiling)
 * @param weights Weight map
 * @param floor Minimum weight (default: 0.05 or 5%)
 * @param ceiling Maximum weight (default: 0.50 or 50%)
 * @returns Constrained weight map
 */
export function applyWeightConstraints(
  weights: WeightMap,
  floor: number = 0.05,
  ceiling: number = 0.50
): WeightMap {
  const constrained: WeightMap = {};

  for (const [provider, weight] of Object.entries(weights)) {
    constrained[provider] = Math.max(floor, Math.min(ceiling, weight));
  }

  // Renormalize after applying constraints
  return normalizeWeights(constrained);
}

/**
 * Softmax weight calculation.
 * Converts scores to probability-like weights where higher scores get more weight.
 * Temperature controls how much to differentiate: lower temp = more extreme weights.
 */
export function calculateSoftmaxWeights(
  scores: Record<string, number>,
  temperature: number = 0.5
): WeightMap {
  const ids = Object.keys(scores);
  if (ids.length === 0) return {};

  // Find max for numerical stability
  const maxScore = Math.max(...Object.values(scores));

  const expValues: Record<string, number> = {};
  let sumExp = 0;

  for (const id of ids) {
    const exp = Math.exp((scores[id] - maxScore) / temperature);
    expValues[id] = exp;
    sumExp += exp;
  }

  const weights: WeightMap = {};
  for (const id of ids) {
    weights[id] = expValues[id] / sumExp;
  }

  return weights;
}

/**
 * Generate dynamic weights based on historical Brier scores.
 * Falls back to equal weights when insufficient history exists.
 *
 * Lower Brier score = better accuracy = higher weight.
 * Uses softmax on negated Brier scores so lower scores get higher weights.
 */
export function getDynamicWeights(
  providerIds: string[],
  accuracyData: ProviderAccuracyRow[],
  minSampleSize: number = 20,
  temperature: number = 0.5,
  weightFloor: number = 0.03,
  weightCeiling: number = 0.40
): WeightMap {
  if (providerIds.length === 0) return {};

  // Build lookup of provider -> accuracy data
  const accuracyMap = new Map<string, ProviderAccuracyRow>();
  for (const row of accuracyData) {
    if (row.provider_type === 'bookmaker') {
      accuracyMap.set(row.provider_name, row);
    }
  }

  // Separate into qualified (enough history) and unqualified
  const qualified: { id: string; brierAvg: number }[] = [];
  const unqualified: string[] = [];

  for (const id of providerIds) {
    const acc = accuracyMap.get(id);
    if (acc && acc.total_predictions >= minSampleSize && acc.brier_score_avg != null) {
      qualified.push({ id, brierAvg: acc.brier_score_avg });
    } else {
      unqualified.push(id);
    }
  }

  // Need at least 3 qualified providers to use dynamic weighting
  if (qualified.length < 3) {
    return generateWeightMapForProviders(providerIds);
  }

  // Calculate softmax weights from negated Brier scores
  // (negate because lower Brier = better, but softmax rewards higher values)
  const negatedScores: Record<string, number> = {};
  for (const q of qualified) {
    negatedScores[q.id] = -q.brierAvg;
  }

  const softmaxWeights = calculateSoftmaxWeights(negatedScores, temperature);

  // Apply floor/ceiling constraints
  const constrained = applyWeightConstraints(softmaxWeights, weightFloor, weightCeiling);

  // Assign unqualified providers the median weight from qualified set
  const qualifiedWeightValues = Object.values(constrained).sort((a, b) => a - b);
  const medianWeight =
    qualifiedWeightValues.length % 2 === 0
      ? (qualifiedWeightValues[qualifiedWeightValues.length / 2 - 1] +
         qualifiedWeightValues[qualifiedWeightValues.length / 2]) /
        2
      : qualifiedWeightValues[Math.floor(qualifiedWeightValues.length / 2)];

  const finalWeights: WeightMap = { ...constrained };
  for (const id of unqualified) {
    finalWeights[id] = medianWeight;
  }

  // Normalize to sum to 1.0
  return normalizeWeights(finalWeights);
}
